helix-rest/src/main/java/org/apache/helix/rest/metadatastore/TrieRoutingData.java

```diff
@@ -0,0 +1,136 @@
+    TrieNode leafNode = findTrieNode(path, true);
+    return leafNode._realmAddress;
+  }
+
+  /**
+   * If findLeafAlongPath is false, then starting from the root node, find the trie node that the
+   * given path is pointing to and return it; raise IllegalArgumentException if the path does
+   * not point to any node. If findLeafAlongPath is true, then starting from the root node, find the
+   * leaf node along the provided path; raise IllegalArgumentException if the path does not
+   * point to any node or if there is no leaf node along the path.
+   * @param path - the path where the search is conducted
+   * @param findLeafAlongPath - whether the search is for a leaf node on the path
+   * @return the node pointed by the path or a leaf node along the path
+   * @throws IllegalArgumentException - when the path points to nothing or when no leaf node is found
+   */
+  private TrieNode findTrieNode(String path, boolean findLeafAlongPath)
+      throws IllegalArgumentException {
+    if (path.equals("/") || path.equals("")) {
+      if (findLeafAlongPath && !_rootNode._isLeaf) {
+        throw new IllegalArgumentException("no leaf node found along the path");
+      }
+      return _rootNode;
+    }
+
+    if (path.substring(0, 1).equals("/")) {
+      path = path.substring(1);
+    }
+    String[] splitPath = path.split("/", 0);
+
+    TrieNode curNode = _rootNode;
+    if (findLeafAlongPath && curNode._isLeaf) {
+      return curNode;
+    }
+    Map<String, TrieNode> curChildren = curNode._children;
+    for (String pathSection : splitPath) {
+      curNode = curChildren.get(pathSection);
+      if (curNode == null) {
+        throw new IllegalArgumentException("the provided path is missing from the trie");
+      }
+      if (findLeafAlongPath && curNode._isLeaf) {
+        return curNode;
+      }
+      curChildren = curNode._children;
+    }
+    if (findLeafAlongPath) {
+      throw new IllegalArgumentException("no leaf node found along the path");
+    }
+    return curNode;
+  }
+
+  /**
+   * Given a trie node, search for all leaf nodes that descend from the given trie node, and add
+   * all complete paths and realm addresses of these leaf nodes to a provided mapping. This method
+   * recursively calls itself.
+   * @param mapping - where the results (complete paths of leaf nodes and their realm addresses)
+   *          are stored
+   * @param curNode - the current trie node where the search starts
+   * @param curPath - the complete path of the current trie node
+   */
+  private static void addAllAddressesToMapping(Map<String, String> mapping, TrieNode curNode,
+      String curPath) {
+    if (curNode._isLeaf) {
+      mapping.put(curPath, curNode._realmAddress);
+      return;
+    }
+
+    for (Map.Entry<String, TrieNode> entry : curNode._children.entrySet()) {
+      addAllAddressesToMapping(mapping, entry.getValue(), curPath + "/" + entry.getKey());
```

Comment:
Let's try to use iterative way instead of recursive call. You may take more memory in memory stack for function recursive call.

Reply:
The recursive calls won't create many call stacks because the depth of the trie is typically limited: based on the current usage, the maximum is 3. The number of call stacks is bounded by the depth of the trie.
In terms of scaling in the future, I do agree that an iterative method is better, however, constructing the full paths for each node is more difficult than necessary. (This is related to your next comment - storing full paths resolves this problem. If you feel strongly about the full-path design, I'm happy to discuss further.)

Reply:
I experimented with "nodes storing full path" and "node storing partial path", and turns out the full path approach makes this method marginally better. I'm therefore adding a field that represents the full path of a trie node.